pytorch_lightning/core/lightning.py

```diff
@@ -255,9 +256,13 @@ def _apply_batch_transfer_handler(
         self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0
     ) -> Any:
         device = device or self.device
-        batch = self.on_before_batch_transfer(batch, dataloader_idx)
-        batch = self.transfer_batch_to_device(batch, device, dataloader_idx)
-        batch = self.on_after_batch_transfer(batch, dataloader_idx)
+        datahook_source = (
+            _DataHookSource(self, None) if self.trainer is None else self.trainer._data_connector._datahook_source
```

Comment:
Is it expected that `self.trainer` would ever be `None` here?

Reply:
if its None, then it would select `_DataHookSoure(self, None)`.

Reply:
I meant if it can happen that this method (`_apply_batch_transfer_handler`) can be called and the trainer reference not be set. Meaning, should it be

```python
assert self.trainer is not None
datahook_source = self.trainer._data_connector._datahook_source
```

(or a ValueError instead of an assertion if it's not just a mypy issue)

Reply:
some tests were failing since this is being used within ModelSummary, `to_onnx`, `to_torchscript` where we don't need a trainer every time.